### PR TITLE
[Quest API] Add Additional XP Events EVENT_AA_EXP_GAIN, EVENT_EXP_GAIN

### PR DESCRIPTION
**Pull request by Valorith**
# Perl
- Add `EVENT_AA_EXP_GAIN`
-> Exports `$aa_exp_gained`
- Add `EVENT_EXP_GAIN`
-> Exports `$exp_gained`
# Lua
- Add `event_aa_exp_gain`
-> Exports `e.aa_exp_gained`
- Add `event_exp_gain`
-->Exports `e.exp_gained`
# Notes
- Allows operators to perform specific actions each time XP or AA XP is gained.

# Testing
- Tested w/ both Perl and Lua and it performs as expected.

### CODE REVIEW

**Comment by Kinglykrab** [comment]
`EVENT_AA_EXP_GAIN` and `EVENT_EXP_GAIN` for verbosity, as well as `aa_exp_gained` and `exp_gained` for the exports. Otherwise looks awesome. :1st_place_medal:

**Comment by Akkadius** [comment]
> `EVENT_AA_EXP_GAIN` and `EVENT_EXP_GAIN` for verbosity, as well as `aa_exp_gained` and `exp_gained` for the exports. Otherwise looks awesome. 1st_place_medal

Agree

File: zone/exp.cpp [hunk 1]
```diff
@@ -801,6 +801,17 @@ void Client::SetEXP(uint64 set_exp, uint64 set_aaxp, bool isrezzexp) {
 		}
 	}
 
+    if (m_pp.exp != set_exp) {
+        const auto xp_value = set_exp - m_pp.exp;
+        const auto export_string = fmt::format("{}",xp_value);
+        parse->EventPlayer(EVENT_XP_GAIN, this,export_string, xp_value);
+    }
+    if (m_pp.expAA != set_aaxp) {
```
Comment (Kinglykrab) [inline zone/exp.cpp:809]: Newline here.

File: zone/exp.cpp [hunk 2]
```diff
@@ -801,6 +801,17 @@ void Client::SetEXP(uint64 set_exp, uint64 set_aaxp, bool isrezzexp) {
 		}
 	}
 
+    if (m_pp.exp != set_exp) {
+        const auto xp_value = set_exp - m_pp.exp;
+        const auto export_string = fmt::format("{}",xp_value);
+        parse->EventPlayer(EVENT_XP_GAIN, this,export_string, xp_value);
```
Comment (Kinglykrab) [inline zone/exp.cpp:807]: Missing a space here.

File: zone/exp.cpp [hunk 3]
```diff
@@ -801,6 +801,17 @@ void Client::SetEXP(uint64 set_exp, uint64 set_aaxp, bool isrezzexp) {
...
+    }
+    if (m_pp.expAA != set_aaxp) {
+        const auto aaxp_value = set_aaxp - m_pp.expAA;
+        const auto export_string = fmt::format("{}",aaxp_value);
+        parse->EventPlayer(EVENT_AAXP_GAIN, this, export_string, aaxp_value);
+    }
+
```
Comment (Kinglykrab) [inline zone/exp.cpp:814]: Don't think you **need** to send it in `data` and `extradata` here, just one or the other is fine.

Perl doesn't really care what you send it, so you can just send `data` and leave `extradata` as `0`.

File: zone/embparser.cpp [hunk 4]
```diff
@@ -1996,6 +1998,16 @@ void PerlembParser::ExportEventVariables(
 			break;
 		}
 
+        case EVENT_AAXP_GAIN: {
+            ExportVar(package_name.c_str(), "aaxp_value", extradata);
+            break;
+        }
+
+        case EVENT_XP_GAIN: {
+            ExportVar(package_name.c_str(), "xp_value", extradata);
```
Comment (Kinglykrab) [inline zone/embparser.cpp:2007]: `data`

File: zone/embparser.cpp [hunk 5]
```diff
@@ -1996,6 +1998,16 @@ void PerlembParser::ExportEventVariables(
 			break;
 		}
 
+        case EVENT_AAXP_GAIN: {
+            ExportVar(package_name.c_str(), "aaxp_value", extradata);
```
Comment (Kinglykrab) [inline zone/embparser.cpp:2002]: `data`

File: zone/exp.cpp (hunk 1 above)
Comment (Valorith) [reply zone/exp.cpp:809]: Updated

File: zone/exp.cpp (hunk 2 above)
Comment (Valorith) [reply zone/exp.cpp:807]: Updated

File: zone/exp.cpp (hunk 3 above)
Comment (Valorith) [reply zone/exp.cpp:814]: Updated

File: zone/embparser.cpp (hunk 4 above)
Comment (Valorith) [reply zone/embparser.cpp:2007]: Updated

File: zone/embparser.cpp (hunk 5 above)
Comment (Valorith) [reply zone/embparser.cpp:2002]: Updated

**Comment by Valorith** [comment]
Updated and successfully tested based on review comments.

**Comment by Kinglykrab** [comment]
> `EVENT_AA_EXP_GAIN` and `EVENT_EXP_GAIN` for verbosity, as well as `aa_exp_gained` and `exp_gained` for the exports. Otherwise looks awesome. 🥇

The exports weren’t addressed.

File: zone/exp.cpp [hunk 6]
```diff
@@ -801,6 +801,18 @@ void Client::SetEXP(uint64 set_exp, uint64 set_aaxp, bool isrezzexp) {
...
 
+    if (m_pp.exp != set_exp) {
+        const auto xp_value = set_exp - m_pp.exp;
+        const auto export_string = fmt::format("{}", xp_value);
+        parse->EventPlayer(EVENT_EXP_GAIN, this, export_string, 0);
+    }
+
+    if (m_pp.expAA != set_aaxp) {
+        const auto aaxp_value = set_aaxp - m_pp.expAA;
+        const auto export_string = fmt::format("{}",aaxp_value);
```
Comment (Kinglykrab) [inline zone/exp.cpp:812]: Space here.

Comment (Valorith) [reply zone/exp.cpp]: Fixed

**Comment by Valorith** [comment]
> > `EVENT_AA_EXP_GAIN` and `EVENT_EXP_GAIN` for verbosity, as well as `aa_exp_gained` and `exp_gained` for the exports. Otherwise looks awesome. 🥇
> 
> The exports weren’t addressed.

Should be good now

File: zone/lua_parser.cpp [hunk 7]
```diff
@@ -148,6 +148,8 @@ const char *LuaEvents[_LargestEventID] = {
 	"event_task_before_update",
 	"event_aa_buy",
 	"event_aa_gain",
+    "event_aa_exp_gain",
```
Comment (Akkadius) [inline zone/lua_parser.cpp:151]: Indent

File: zone/lua_parser.cpp [hunk 8]
```diff
@@ -273,6 +275,8 @@ LuaParser::LuaParser() {
 	PlayerArgumentDispatch[EVENT_INSPECT]                    = handle_player_inspect;
 	PlayerArgumentDispatch[EVENT_AA_BUY]                     = handle_player_aa_buy;
 	PlayerArgumentDispatch[EVENT_AA_GAIN]                    = handle_player_aa_gain;
+    PlayerArgumentDispatch[EVENT_AA_EXP_GAIN]                = handle_player_aa_exp_gain;
```
Comment (Akkadius) [inline zone/lua_parser.cpp:278]: Indent

File: zone/exp.cpp [hunk 9]
```diff
@@ -801,6 +801,18 @@ void Client::SetEXP(uint64 set_exp, uint64 set_aaxp, bool isrezzexp) {
 		}
 	}
 
+    if (m_pp.exp != set_exp) {
+        const auto exp_value = set_exp - m_pp.exp;
+        const auto export_string = fmt::format("{}", exp_value);
+        parse->EventPlayer(EVENT_EXP_GAIN, this, export_string, 0);
+    }
+
+    if (m_pp.expAA != set_aaxp) {
```
Comment (Akkadius) [inline zone/exp.cpp:810]: These both should be gated with `parse->PlayerHasQuestSub(`

File: zone/lua_parser.cpp (hunk 7 above)
Comment (Valorith) [reply zone/lua_parser.cpp]: Fixed

File: zone/lua_parser.cpp (hunk 8 above)
Comment (Valorith) [reply zone/lua_parser.cpp:278]: Fixed

File: zone/lua_general.cpp [hunk 10]
```diff
@@ -4730,7 +4730,9 @@ luabind::scope lua_register_events() {
 			luabind::value("inspect", static_cast<int>(EVENT_INSPECT)),
 			luabind::value("task_before_update", static_cast<int>(EVENT_TASK_BEFORE_UPDATE)),
 			luabind::value("aa_buy", static_cast<int>(EVENT_AA_BUY)),
-			luabind::value("aa_gain", static_cast<int>(EVENT_AA_GAIN)),
+			luabind::value("aa_gained", static_cast<int>(EVENT_AA_GAIN)),
+            luabind::value("aa_exp_gained", static_cast<int>(EVENT_AA_EXP_GAIN)),
```
Comment (Akkadius) [inline zone/lua_general.cpp:4734]: Indent

File: zone/exp.cpp (hunk 9 above)
Comment (Valorith) [reply zone/exp.cpp:810]: Added

File: zone/lua_general.cpp (hunk 10 above)
Comment (Valorith) [reply zone/lua_general.cpp:4734]: Fixed

**Comment by Valorith** [comment]
Now I know to pay attention to indenting. I was thinking if it lined up on the screen it was good, but now I see indenting matters.

File: zone/embparser.cpp [hunk 11]
```diff
@@ -1996,6 +1998,16 @@ void PerlembParser::ExportEventVariables(
 			break;
 		}
 
+        case EVENT_AA_EXP_GAIN: {
```
Comment (Akkadius) [inline zone/embparser.cpp:2001]: Formatting is off here with the rest of the switch

File: zone/exp.cpp [hunk 12]
```diff
@@ -801,6 +801,18 @@ void Client::SetEXP(uint64 set_exp, uint64 set_aaxp, bool isrezzexp) {
 		}
 	}
 
+    if (parse->PlayerHasQuestSub(EVENT_EXP_GAIN) && m_pp.exp != set_exp) {
```
Comment (Akkadius) [inline zone/exp.cpp:804]: Indent is now off here as well, tabs / spaces

File: zone/lua_parser_events.cpp [hunk 13]
```diff
@@ -1163,6 +1163,30 @@ void handle_player_aa_gain(
...
+
+void handle_player_exp_gain(
+        QuestInterface *parse,
+        lua_State* L,
+        Client* client,
+        std::string data,
+        uint32 extra_data,
+        std::vector<std::any> *extra_pointers
+) {
+    lua_pushinteger(L, std::stoull(data));
```
Comment (Akkadius) [inline zone/lua_parser_events.cpp:1186]: Indentation is weird here as well

File: zone/embparser.cpp (hunk 11 above)
Comment (Akkadius) [reply zone/embparser.cpp:2001]: This is still off

**Review by Kinglykrab — APPROVED** [review]

**Pull request merged**